recon/src/protocol.rs

```diff
@@ -142,275 +139,360 @@ where
-            let stream = self.role.stream();
-            if let Some(message) = stream.try_next().await? {
-                self.metrics.record(&MessageRecv(&message));
-                self.handle_incoming(message)
-                    .await
-                    .context("handle incoming")?;
+#[instrument(skip_all, fields(sync_id))]
+async fn write<M, S>(
+    mut sync_id: Option<String>,
+    sink: S,
+    mut to_writer_rx: mpsc::Receiver<ToWriter<M>>,
+    init: Option<M>,
+    finish: Option<M>,
+    wip_limit: usize,
+    metrics: Metrics,
+) -> Result<()>
+where
+    S: Sink<ReconMessage<M>, Error = anyhow::Error>,
+    MessageLabels: for<'a> From<&'a M>,
+    M: std::fmt::Debug,
+{
+    if let Some(sync_id) = &sync_id {
+        // Record sync_id on the tracing span
+        tracing::Span::current().record("sync_id", sync_id);
+    }
+    pin_mut!(sink);
+
+    // Use a stack so we get depth first traversal
+    let mut message_stack = Vec::new();
+    // Track in progress
+    let mut in_progress = 0;
+
+    if let Some(init) = init {
+        trace!(?init, "sending init");
+        metrics.record(&MessageSent(&init));
+        sink.send(ReconMessage::new(sync_id.clone(), init)).await?;
+    }
+
+    // Write messages to the remote until there are no more messages to send.
+    while let Some(action) = to_writer_rx.recv().await {
+        metrics.record(&ProtocolWriteLoop);
+        match action {
+            ToWriter::SyncId(remote_sync_id) => {
+                // Record sync_id on the tracing span
+                tracing::Span::current().record("sync_id", &remote_sync_id);
+                sync_id = Some(remote_sync_id);
             }
-
-            self.role.each().await?;
-
-            if self.role.is_done() && self.remote_done {
+            ToWriter::SendAll(messages) => {
+                sink.send_all(&mut messages.map(|msg| {
+                    msg.map(|message| {
+                        metrics.record(&MessageSent(&message));
+                        trace!(?message, "sending message");
+                        ReconMessage::new(sync_id.clone(), message)
+                    })
+                }))
+                .await?;
+            }
+            ToWriter::WIPCompleted(count) => {
+                in_progress -= count;
+            }
+            ToWriter::SendWIPLimited(mut messages) => {
+                // Send all message that fit within the work in progress limit.
+                if in_progress < wip_limit {
+                    let capacity = wip_limit - in_progress;
+                    let mut sent = false;
+                    for message in messages.drain(..std::cmp::min(messages.len(), capacity)) {
+                        in_progress += 1;
+                        sent = true;
+                        trace!(?message, "feeding message");
+                        metrics.record(&MessageSent(&message));
+                        sink.feed(ReconMessage::new(sync_id.clone(), message))
```

Comment:
I'm curious how feed/flush vs send/send_all behaves. This seems like the natural way to do this rather than keep things in memory, but I'd rather have a synchronous feed API and then async flush().. if I'm giving up control I might as well `send()`, no?

Anyway, before this seemed to be related to the slowing message exchange in conversations. There definitely was a flush after X setting coming into play in our case, but I guess this is a long winded way of saying maybe we should just call send (or map and send_all if we can)? 

If we feed and it's full, we might wait a while without actually being able to send things... I guess we could flush every X items or something, and maybe it's silly to worry about now. 

Basically, my opinion is that using the feed API makes sense, but given how abstracted everything is through libp2p/os/etc to the actual I/O, I think it might be better to just hit send and let the lower layers buffer things and keep things flowing steadily.

Reply:
> Basically, my opinion is that using the feed API makes sense, but given how abstracted everything is through libp2p/os/etc to the actual I/O, I think it might be better to just hit send and let the lower layers buffer things and keep things flowing steadily.

My understanding of the API is that using `send` prevents any lower layers from buffering as it calls flush down the layers of abstraction ensuring the message has been sent all the way to the remote node.

From the docs on feed:

```
A future that completes after the given item has been received
by the sink.

Unlike `send`, the returned future does not flush the sink.
It is the caller's responsibility to ensure all pending items
are processed, which can be done via `flush` or `close`.
```

From the docs on send:

```
A future that completes after the given item has been fully processed
into the sink, including flushing.

Note that, **because of the flushing requirement, it is usually better
to batch together items to send via `feed` or `send_all`,
rather than flushing between each item.**

```

As recommended when we use `send_all` or `feed` depending on if we have a stream of messages or just a set of messages in memory. Additionally we always call `flush` after doing a batch of `feed` calls so its not possible that a message is not flushed and therefore stuck in pending land. 

This differs from how we did this previously because we used feed on a counter so its was possible a message landed on the wrong side of the count and wasn't flushed to the other node. This cannot happen with this code change as we always flush after calling feed.

Reply:
We discussed and no changes needed 😄 and I agree the loop/flush every time now versus the count before is good. I just wanted to call out the sink handling as an area we might want to keep in mind, but intuitively, and according to the docs, using feed like this is the right thing to do.